### ISSUE TITLE
fp16 performance issues

### ISSUE THREAD

**Issue by fshi98** [ROCmSoftwarePlatform/tensorflow-upstream#256]
I have tested with vega64, ubuntu18.04, ROCm1.9.2, tf1.12:
1 resnet50: python tf_cnn_benchmarks.py --num_gpus=1 --batch_size=64 --model=resnet50
  1080ti: 212 images/sec  (278 fp16)     
  vega64: 191 images/sec (190 fp16)
2 resnet101: python tf_cnn_benchmarks.py --num_gpus=1 --batch_size=32 --model=resnet101
  1080ti: 121.14 images/sec (168 fp16)     
  vega64: 101.15 images/sec (93 fp16), if fp16, --batch_size could be 64, while fp32, 64 will crash
3. inception3: python tf_cnn_benchmarks.py --num_gpus=1 --batch_size=64 --model=inception3
  1080ti: 140.08 images/sec (166 fp16)    
  vega64: 99.02 images/sec (50 fp16)

4 mobilenet: python tf_cnn_benchmarks.py --num_gpus=1 --batch_size=64 --model=mobilenet
  1080ti: 2865 images/sec     
  vega64: 462 images/sec

The nv gtx1080 ti was tested on another machine with cuda10, ubuntu 18.04.

There are two values didn't add up:
1. for mobilenet, the 1080ti result doesn't make sense. 
2. i also tested with --use_fp16, which gives  fair amount of speedup for 1080ti. However, for vega64, it ends up slower in all tests if using --use_fp16. This is especially true for inception3.  

Considering vega64 supports native half precision and fp16 should be a good selling point for amd vega. how is it slower if using fp16? I guess this is probably due to software support, especially ROCm. Can anyone please test it with --use_fp16 and see if having similar results. 

_Originally posted by @fshi98 in https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/issues/173#issuecomment-443325358_

**Comment by dagamayank** [comment]
@fshi98 thanks for the issue. TensorFlow and MIOpen are not optimized for `fp16` on Vega64. They are only for functional enablement. Future GPUs from AMD like the recently announced MI60 will be `fp16` optimized.

Please report feedback if you see issues with `fp32` on Vega64.

**Comment by witeko** [comment]
@dagamayank will the radeon vii be optimized for fp16? Pls reply when You can. :) (i know it's not out yet)

**Comment by Mandrewoid** [comment]
@dagamayank I am a little confused with your statement.  "Future GPUs from AMD like the recently announced MI60 will be fp16 optimized." The previous statement implied the issue was in software(Namely MIOpen and TensorFlow) but then your next sentence suggests it is hardware related.
Vega does have double FLOPS for fp16, and MI60 is also based on vega. This seems like an artificial limitation?

**Comment by ubombi** [comment]
Is Vega64/FE hardware capable of fast FP16?
 
TensorFlow and MIOpen are _not optimized **yet**_ OR  _**will newer be** optimized_ for `fp16`?

**Comment by dagamayank** [comment]
> Is Vega64/FE hardware capable of fast FP16?

What is your use-case, inference or training?

**Comment by Mandrewoid** [comment]
> > Is Vega64/FE hardware capable of fast FP16?
> 
> What is your use-case, inference or training?

If  it isn't too much to ask, an answer for both would be nice.

**Comment by seesturm** [comment]
Here is my speculation for the claim of Vega64 being "not optimized for fp16". It is the lack of the new "deep-learning" v_dot2 instruction which comes with Radeon 7. Altough Vega64 has has the v_mac_f16 instruction, it has only a fp16 output instead of fp32 with v_dot2. So I guess when (matrix?) multiplication is performed using v_mac_f16 the loss of precision makes the result of the accumulated rounding error in the sum too big for deep learning.
If this is true it would still allow to implement using v_mac_f16 when one accepts the loss of precision. But I have no idea if this would be useful for deep learning at all.
